gremlin-driver/src/test/java/org/apache/tinkerpop/gremlin/driver/WebSocketClientBehaviorIntegrateTest.java

```diff
@@ -271,4 +276,28 @@ public void shouldNotCreateReplacementConnectionWhenClientClosesConnection() thr
                         .filter(str -> str.contains("Considering new connection on"))
                         .count());
     }
-}
\ No newline at end of file
+
+    /**
+     * (TINKERPOP-2814) Tests to make sure that the SSL handshake is now capped by connectionSetupTimeoutMillis and not
+     * the default Netty SSL handshake timeout of 10,000ms.
+     */
+    @Test
+    public void shouldAttemptHandshakeForLongerThanDefaultNettySslHandshakeTimeout() {
+        final Cluster cluster = Cluster.build("localhost").port(SimpleSocketServer.PORT)
+                .minConnectionPoolSize(1)
+                .maxConnectionPoolSize(1)
+                .connectionSetupTimeoutMillis(20000) // needs to be larger than 10,000ms.
+                .enableSsl(true)
+                .create();
+
+        final Client.ClusteredClient client = cluster.connect();
+        final long start = System.currentTimeMillis();
+
+        try {
+            client.submit("1");
+        } finally {
+            // Test against 15,000ms which should give a big enough buffer to avoid timing issues.
+            assertTrue(System.currentTimeMillis() - start > 15000);
```

Comment:
If I understand this correctly, we expect the connection setup to fail and we verify that the failure is after the threshold of ssl handshake timeout, hence proving that the timeout is bound by the overall connection setup limit and not by ssl. Is that right?

Could we also add an assertion that we expect the connection setup to fail? I know that it is not what this test is checking but it will help guard the purpose of this test is future cases. Also let's validate that the correct error message is being printed and we haven't regressed in the aspect of providing the detailed/correct error log.

Reply:
Yes, your understanding of the test is correct. Sure, I'll add those assertions.